### PR TITLE
Add community confidence voting

### PR DESCRIPTION
**Pull request by hgtonight**
Round 2

This adds a community confidence rating to addons.

It is a work in progress.

* [x] Detect Vanilla (core) versions
* [x] Display community rating of plugins in addon page
* [x] Thumb up/down voting

### CODE REVIEW

**Comment by hgtonight** [comment]
I think this is enough to get started.

**Comment by hgtonight** [comment]
Closes #6

File: applications/addons/controllers/class.addoncontroller.php [hunk 1]
```diff
@@ -93,6 +95,72 @@ public function index($ID = '') {
...
+            throw permissionException('@You need to be logged in to vote.');
+        }
+                
+        $addon = $this->AddonModel->getVersion($addonVersionID);
+        if(!$addon) {
+            throw notFoundException('Addon');
+        }
+        
+        $currentVote = $this->ConfidenceModel->getConfidenceVote($session->UserID, $addon['AddonVersionID'], $coreVersionID);
+        if($currentVote === false) {
```
Comment (linc) [inline applications/addons/controllers/class.addoncontroller.php]: Don't forget your spaces after `if`s.

**Comment by linc** [comment]
Definitely needs inline docs.

**Comment by hgtonight** [comment]
Added some doc blocks and fixed the spaces.

**Comment by hgtonight** [comment]
I did not mean to set the estimate.

File: applications/addons/controllers/class.addoncontroller.php [hunk 2]
```diff
@@ -94,6 +96,109 @@ public function index($ID = '') {
...
+    
+    /**
+     * Mark the current user as saying the addon works with a specific version
+     * of Vanilla
+     * 
+     * @param int $addonVersionID
+     * @param int $coreVersionID default to the latest verion of Vanilla
+     */
+    public function works($addonVersionID, $coreVersionID = false) {
+        $this->updateVote($addonVersionID, $coreVersionID, 1);
```
Comment (linc) [inline applications/addons/controllers/class.addoncontroller.php]: This endpoint needs CSRF protection - respond only to `POST` and check the transientkey.

File: applications/addons/controllers/class.addoncontroller.php [hunk 3]
```diff
@@ -94,6 +96,109 @@ public function index($ID = '') {
...
+     * 
+     * @param int $addonVersionID
+     * @param int $coreVersionID
+     * @param int $weight
+     * @throws NotFoundException, PermissionException
+     */
+    private function updateVote($addonVersionID, $coreVersionID, $weight) {
+        $session = Gdn::Session();
+        if (!$session->isValid()) {
+            throw permissionException('@You need to be logged in to vote.');
```
Comment (linc) [inline applications/addons/controllers/class.addoncontroller.php]: There's a translation already for `'You must be signed in.'` which I recommend using here.

File: applications/addons/controllers/class.addoncontroller.php [hunk 4]
```diff
@@ -94,6 +96,109 @@ public function index($ID = '') {
...
+        
+        $currentVote = $this->ConfidenceModel->getConfidenceVote($session->UserID, $addon['AddonVersionID'], $coreVersionID);
+        if ($currentVote === false) {
+            $this->ConfidenceModel->insert([
+                'AddonVersionID' => $addon['AddonVersionID'],
+                'UserID' => $session->UserID,
+                'Weight' => $weight]);
+        }
+        else {
+            if ($currentVote->Weight != $weight) {
```
Comment (linc) [inline applications/addons/controllers/class.addoncontroller.php]: You don't like `elseif`? :o)

File: applications/addons/models/class.confidencemodel.php [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+     * Loads the confidence summary of the specified addon version against the 
+     * specified core version.
+     * 
+     * @param int $AddonVersionID
+     * @param string $DataType DATASET_TYPE_ARRAY or DATASET_TYPE_OBJECT
+     * @return array|stdClass matching $DataType
+     */
+    public function getID($AddonVersionID, $coreVersionID = false, $DataType = DATASET_TYPE_ARRAY) {
+        $coreVersion = $this->checkCoreVersion($coreVersionID);
+        $Confidence = $this->SQL
```
Comment (linc) [inline applications/addons/models/class.confidencemodel.php]: I'd be more OK with using the old coding standards if you weren't randomly switching between casings. :)

File: applications/addons/settings/structure.php [hunk 6]
```diff
@@ -58,55 +58,74 @@
...
+$Construct->Table('AddonComment')
+    ->PrimaryKey('AddonCommentID')
+    ->Column('AddonID', 'int', FALSE, 'key')
+    ->Column('InsertUserID', 'int', FALSE, 'key')
+    ->Column('Body', 'text')
+    ->Column('Format', 'varchar(20)', TRUE)
+    ->Column('DateInserted', 'datetime')
+    ->Set($Explicit, $Drop);
+*/
+
```
Comment (linc) [inline applications/addons/settings/structure.php]: Why is this entire file reverted?

File: applications/addons/views/addon/addon.php [hunk 7]
```diff
@@ -52,6 +52,55 @@
...
+            <div class="Box AddonBox ConfidenceBox">
+                <?php
+                $addonVersionID = $this->Data('AddonVersionID');
+                $confidence = $this->ConfidenceModel->getID($addonVersionID, false, DATASET_TYPE_OBJECT);
+                $coreVersion = $this->ConfidenceModel->getCoreVersion();
+                
+                echo wrap(sprintf(t('Vanilla %s Compatibility'), $coreVersion->Version), 'h3');
+                
+                if (!$confidence) {
+                    echo Wrap(sprite('Bandaid', 'BigSprite', 'Unsure') . T('The community has said nothing.'), 'p');
```
Comment (linc) [inline applications/addons/views/addon/addon.php]: Again - 2 function calls in 4 lines, one caps one lowercase is a bit much.

**Comment by linc** [comment]
I made some inline notes which I'll summarize:

* New endpoints need CSRF protection.
* Prefer existing translations over creating similar new ones.
* Coding standards are switching around randomly, especially casing. Also note that docblock sentences should all end with a period consistently; I noticed that's pretty random too.

File: applications/addons/controllers/class.addoncontroller.php (hunk 4 above)
Comment (hgtonight) [inline applications/addons/controllers/class.addoncontroller.php]: Love me them indents ;)

File: applications/addons/controllers/class.addoncontroller.php (hunk 2 above)
Comment (hgtonight) [inline applications/addons/controllers/class.addoncontroller.php]: Added the check to the updateVote() method;

**Comment by hgtonight** [comment]
I believe I have fixed all your comments.

**Comment by linc** [comment]
I made a few changes:

* Moved it below the download box
* Removed it from locales, which are far less version-dependent.
* Tweaked the type check to use `strtolower()` which is a better standard check to use.

**Comment by linc** [comment]
The sprites are maybe a bit much but we can iterate on the design a bit. ;)

**Pull request merged**